`crux` fails to build with `aeson-2.0.*`

Trying to build `crux` with `aeson-2.0.0.0` or later will fail with the following errors:

<details>

```
[16 of 21] Compiling Crux.SVCOMP      ( src/Crux/SVCOMP.hs, /home/rscott/Documents/Hacking/Haskell/sandbox/crucible/dist-newstyle/build/x86_64-linux/ghc-8.10.4/crux-0.5.0.99/build/Crux/SVCOMP.o, /home/rscott/Documents/Hacking/Haskell/sandbox/crucible/dist-newstyle/build/x86_64-linux/ghc-8.10.4/crux-0.5.0.99/build/Crux/SVCOMP.dyn_o )

src/Crux/SVCOMP.hs:230:38: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k0 Yaml.Value’
      Expected type: HM.HashMap k0 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "format_version" o
      In a stmt of a 'do' block:
        _ <- case HM.lookup "format_version" o of
               Just (Yaml.String "2.0") -> return ()
               _ -> fail
                      $ unwords
                          ["Expected verification task version 2.0 while parsing", show fp]
    |
230 |   do case HM.lookup "format_version" o of
    |                                      ^

src/Crux/SVCOMP.hs:234:41: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k1 Yaml.Value’
      Expected type: HM.HashMap k1 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "input_files" o
      In a stmt of a 'do' block:
        fs <- case HM.lookup "input_files" o of
                Just v -> getStrs v
                Nothing
                  -> fail
                       $ unwords ["No 'input_files' section while parsing", show fp]
    |
234 |      fs <- case HM.lookup "input_files" o of
    |                                         ^

src/Crux/SVCOMP.hs:238:40: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k2 Yaml.Value’
      Expected type: HM.HashMap k2 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "properties" o
      In a stmt of a 'do' block:
        ps <- case HM.lookup "properties" o of
                Just v -> getProps v
                Nothing
                  -> fail
                       $ unwords ["No 'properties' section while parsing", show fp]
    |
238 |      ps <- case HM.lookup "properties" o of
    |                                        ^

src/Crux/SVCOMP.hs:242:39: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k3 Yaml.Value’
      Expected type: HM.HashMap k3 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "options" o
      In a stmt of a 'do' block:
        lang <- case HM.lookup "options" o of
                  Just v -> getLang v
                  Nothing
                    -> fail $ unwords ["No 'options' section while parsing", show fp]
    |
242 |      lang <- case HM.lookup "options" o of
    |                                       ^

src/Crux/SVCOMP.hs:269:47: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k6 Yaml.Value’
      Expected type: HM.HashMap k6 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "property_file" o
      In a stmt of a 'do' block:
        propf <- case HM.lookup "property_file" o of
                   Just (Yaml.String f) -> return (Text.unpack f)
                   _ -> fail
                          $ unwords
                              ["expected string value in 'property_file' while parsing", show fp]
    |
269 |    do propf <- case HM.lookup "property_file" o of
    |                                               ^

src/Crux/SVCOMP.hs:275:52: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k7 Yaml.Value’
      Expected type: HM.HashMap k7 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "expected_verdict" o
      In a stmt of a 'do' block:
        verdict <- case HM.lookup "expected_verdict" o of
                     Nothing -> return Nothing
                     Just (Yaml.Bool b) -> return (Just b)
                     Just (Yaml.String "true") -> return (Just True)
                     Just (Yaml.String "false") -> return (Just False)
                     Just _
                       -> fail
                            $ unwords
                                ["expected boolean value in 'expected_verdict' while parsing",
                                 show fp]
    |
275 |       verdict <- case HM.lookup "expected_verdict" o of
    |                                                    ^

src/Crux/SVCOMP.hs:290:30: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k4 Yaml.Value’
      Expected type: HM.HashMap k4 Yaml.Value
        Actual type: Yaml.Object
    • In the second argument of ‘HM.lookup’, namely ‘o’
      In the expression: HM.lookup "language" o
      In the expression:
        case HM.lookup "language" o of
          Just (Yaml.String "C") -> getCDataModel o
          Just (Yaml.String "Java") -> return Java
          _ -> fail
                 $ unwords
                     ["expected 'C' or 'Java' in 'language' while parsing", show fp]
    |
290 |    case HM.lookup "language" o of
    |                              ^

src/Crux/SVCOMP.hs:291:49: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Yaml.Value’
                     with ‘HM.HashMap k5 Yaml.Value’
      Expected type: HM.HashMap k5 Yaml.Value
        Actual type: Yaml.Object
    • In the first argument of ‘getCDataModel’, namely ‘o’
      In the expression: getCDataModel o
      In a case alternative: Just (Yaml.String "C") -> getCDataModel o
    |
291 |      Just (Yaml.String "C")    -> getCDataModel o
    |                                                 ^
```
</details>

This is because `aeson` replaces the `HashMap` field of `Object` with a [`KeyMap` type](https://hackage.haskell.org/package/aeson-2.0.1.0/docs/Data-Aeson-KeyMap.html#t:KeyMap). We should adjust the code accordingly.

--- a/crux-mir/crux-mir.cabal
+++ b/crux-mir/crux-mir.cabal
@@ -20,7 +20,7 @@ extra-source-files:  README.md
 library
   default-language: Haskell2010
   build-depends: base >= 4.7 && < 5,
-                 aeson,
+                 aeson < 2.1,
                  ansi-terminal,
                  bv-sized,
                  bytestring,

--- a/crux-mir/src/Mir/JSON.hs
+++ b/crux-mir/src/Mir/JSON.hs
@@ -4,6 +4,7 @@ Description      : JSON to Mir AST parser
 License          : BSD3
 -}
 
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
@@ -11,7 +12,6 @@ module Mir.JSON where
 
 import Data.Aeson
 import qualified Data.Aeson.Types  as Aeson
-import qualified Data.HashMap.Lazy as HML
 import qualified Data.Map.Strict   as Map
 import qualified Data.Scientific   as Scientific
 
@@ -26,7 +26,14 @@ import Data.List
 import qualified Data.Vector as V
 import Control.Lens((^.),(&),(.~))
 
-import Mir.DefId 
+#if MIN_VERSION_aeson(2,0,0)
+import Data.Aeson.Key (Key)
+import qualified Data.Aeson.KeyMap as KM
+#else
+import qualified Data.HashMap.Lazy as HML
+#endif
+
+import Mir.DefId
 import Mir.Mir
 
 import Debug.Trace
@@ -39,7 +46,7 @@ import Debug.Trace
 
 instance FromJSON BaseSize where
     parseJSON = withObject "BaseSize" $
-                \t -> case HML.lookup "kind" t of
+                \t -> case lookupKM "kind" t of
                         Just (String "Usize") -> pure USize
                         Just (String "U8") -> pure B8
                         Just (String "U16") -> pure B16
@@ -55,7 +62,7 @@ instance FromJSON BaseSize where
                         sz -> fail $ "unknown base size: " ++ show sz
 
 instance FromJSON FloatKind where
-    parseJSON = withObject "FloatKind" $ \t -> case HML.lookup "kind" t of
+    parseJSON = withObject "FloatKind" $ \t -> case lookupKM "kind" t of
                                                  Just (String "F32") -> pure F32
                                                  Just (String "F64") -> pure F64
                                                  sz -> fail $ "unknown float type: " ++ show sz
@@ -76,7 +83,7 @@ instance FromJSON Ty where
 newtype InlineTy = InlineTy { getInlineTy :: Ty }
 
 instance FromJSON InlineTy where
-    parseJSON = withObject "InlineTy" $ \v -> InlineTy <$> case HML.lookup "kind" v of
+    parseJSON = withObject "InlineTy" $ \v -> InlineTy <$> case lookupKM "kind" v of
       Just (String "Bool") -> pure TyBool
       Just (String "Char") -> pure TyChar
       Just (String "Int") -> TyInt <$> v .: "intkind"
@@ -107,7 +114,7 @@ instance FromJSON NamedTy where
         NamedTy <$> v .: "name" <*> (getInlineTy <$> v .: "ty")
 
 instance FromJSON Instance where
-    parseJSON = withObject "Instance" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Instance" $ \v -> case lookupKM "kind" v of
         Just (String "Item") -> Instance IkItem
             <$> v .: "def_id" <*> v .: "substs"
         Just (String "Intrinsic") -> Instance IkIntrinsic
@@ -135,7 +142,7 @@ instance FromJSON FnSig where
                <*> v .: "output"
                <*> v .: "abi"
                <*> spread
-               
+
 instance FromJSON Adt where
     parseJSON = withObject "Adt" $ \v -> Adt
         <$> v .: "name"
@@ -154,13 +161,13 @@ instance FromJSON AdtKind where
         _ -> fail $ "unsupported adt kind " ++ show x
 
 instance FromJSON VariantDiscr where
-    parseJSON = withObject "VariantDiscr" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "VariantDiscr" $ \v -> case lookupKM "kind" v of
                                                     Just (String "Explicit") -> Explicit <$> v .: "name"
                                                     Just (String "Relative") -> Relative <$> v .: "index"
                                                     _ -> fail "unspported variant discriminator"
 
 instance FromJSON CtorKind where
-    parseJSON = withObject "CtorKind" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "CtorKind" $ \v -> case lookupKM "kind" v of
                                                 Just (String "Fn") -> pure FnKind
                                                 Just (String "Const") -> pure ConstKind
                                                 Just (String "Fictive") -> pure FictiveKind
@@ -172,7 +179,7 @@ instance FromJSON Field where
     parseJSON = withObject "Field" $ \v -> Field <$> v .: "name" <*> v .: "ty"
 
 instance FromJSON Mutability where
-    parseJSON = withObject "Mutability" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Mutability" $ \v -> case lookupKM "kind" v of
                                                 Just (String "MutMutable") -> pure Mut
                                                 Just (String "Mut") -> pure Mut
                                                 Just (String "MutImmutable") -> pure Immut
@@ -220,17 +227,17 @@ instance FromJSON Fn where
       Fn
         <$> v .: "name"
         <*> return args
-        <*> sig        
+        <*> sig
         <*> v .: "body"
 
 instance FromJSON Abi where
-    parseJSON = withObject "Abi" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Abi" $ \v -> case lookupKM "kind" v of
         Just (String "Rust") -> pure RustAbi
         Just (String "RustIntrinsic") -> pure RustIntrinsic
         Just (String "RustCall") -> pure RustCall
         Just (String _) -> pure OtherAbi
         x -> fail $ "bad abi: " ++ show x
-        
+
 instance FromJSON BasicBlock where
     parseJSON = withObject "BasicBlock" $ \v -> BasicBlock
         <$> v .: "blockid"
@@ -242,7 +249,7 @@ instance FromJSON BasicBlockData where
         <*> v .: "terminator"
 
 instance FromJSON Statement where
-    parseJSON = withObject "Statement" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Statement" $ \v -> case lookupKM "kind" v of
                              Just (String "Assign") ->  Assign <$> v.: "lhs" <*> v .: "rhs" <*> v .: "pos"
                              Just (String "SetDiscriminant") -> SetDiscriminant <$> v .: "lvalue" <*> v .: "variant_index"
                              Just (String "StorageLive") -> StorageLive <$> v .: "slvar"
@@ -259,7 +266,7 @@ instance FromJSON RustcPlace where
 
 instance FromJSON PlaceElem where
     parseJSON = withObject "PlaceElem" $ \v ->
-      case HML.lookup "kind" v of
+      case lookupKM "kind" v of
         Just (String "Deref") -> pure Deref
         Just (String "Field") -> PField <$> v .: "field" <*> v .: "ty"
         Just (String "Index") -> Index <$> v .: "op"
@@ -274,7 +281,7 @@ instance FromJSON Lvalue where
         convert (RustcPlace base elems) = foldl LProj (LBase base) elems
 
 instance FromJSON Rvalue where
-    parseJSON = withObject "Rvalue" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Rvalue" $ \v -> case lookupKM "kind" v of
                                               Just (String "Use") -> Use <$> v .: "usevar"
                                               Just (String "Repeat") -> Repeat <$> v .: "op" <*> v .: "len"
                                               Just (String "Ref") ->  Ref <$> v .: "borrowkind" <*> v .: "refvar" <*> v .: "region"
@@ -290,7 +297,7 @@ instance FromJSON Rvalue where
                                               k -> fail $ "unsupported RValue " ++ show k
 
 instance FromJSON Terminator where
-    parseJSON = withObject "Terminator" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Terminator" $ \v -> case lookupKM "kind" v of
                                                   Just (String "Goto") -> Goto <$> v .: "target"
                                                   Just (String "SwitchInt") ->
                                                     let  q :: Aeson.Parser [Maybe Integer]
@@ -310,14 +317,14 @@ instance FromJSON Terminator where
                                                   k -> fail $ "unsupported terminator" ++ show k
 
 instance FromJSON Operand where
-    parseJSON = withObject "Operand" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "Operand" $ \v -> case lookupKM "kind" v of
                                                Just (String "Move") -> Move <$> v .: "data"
-                                               Just (String "Copy") -> Copy <$> v .: "data"  
+                                               Just (String "Copy") -> Copy <$> v .: "data"
                                                Just (String "Constant") -> OpConstant <$> v .: "data"
                                                x -> fail ("base operand: " ++ show x)
 
 instance FromJSON NullOp where
-    parseJSON = withObject "NullOp" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "NullOp" $ \v -> case lookupKM "kind" v of
                                              Just (String "SizeOf") -> pure SizeOf
                                              Just (String "Box") -> pure Box
                                              x -> fail ("bad nullOp: " ++ show x)
@@ -329,18 +336,18 @@ instance FromJSON BorrowKind where
       else if t == "Mut"    then pure Mutable
       -- s can be followed by "{ allow_two_phase_borrow: true }"
       else if T.isPrefixOf "Mut" t then pure Mutable
-      else fail ("bad borrowKind: " ++ show t) 
-       
+      else fail ("bad borrowKind: " ++ show t)
 
 
-      
+
+
 instance FromJSON UnOp where
-    parseJSON = withObject "UnOp" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "UnOp" $ \v -> case lookupKM "kind" v of
                                              Just (String "Not") -> pure Not
                                              Just (String "Neg") -> pure Neg
                                              x -> fail ("bad unOp: " ++ show x)
 instance FromJSON BinOp where
-    parseJSON = withObject "BinOp" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "BinOp" $ \v -> case lookupKM "kind" v of
                                              Just (String "Add") -> pure Add
                                              Just (String "Sub") -> pure Sub
                                              Just (String "Mul") -> pure Mul
@@ -370,7 +377,7 @@ instance FromJSON Vtable where
         Vtable <$> v .: "name" <*> v .: "items"
 
 instance FromJSON CastKind where
-    parseJSON = withObject "CastKind" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "CastKind" $ \v -> case lookupKM "kind" v of
                                                Just (String "Misc") -> pure Misc
                                                Just (String "Pointer(ReifyFnPointer)") -> pure ReifyFnPointer
                                                Just (String "Pointer(ClosureFnPointer(Normal))") -> pure ClosureFnPointer
@@ -404,7 +411,7 @@ newtype RustcRenderedConst = RustcRenderedConst ConstVal
 
 instance FromJSON RustcRenderedConst where
     parseJSON = withObject "RenderedConst" $ \v ->
-      RustcRenderedConst <$> case HML.lookup "kind" v of
+      RustcRenderedConst <$> case lookupKM "kind" v of
         Just (String "isize") -> do
             val <- convertIntegerText =<< v .: "val"
             pure $ ConstInt $ Isize val
@@ -498,7 +505,7 @@ convertIntegerText t = do
 
 
 instance FromJSON AggregateKind where
-    parseJSON = withObject "AggregateKind" $ \v -> case HML.lookup "kind" v of
+    parseJSON = withObject "AggregateKind" $ \v -> case lookupKM "kind" v of
                                                      Just (String "Array") -> AKArray <$> v .: "ty"
                                                      Just (String "Tuple") -> pure AKTuple
                                                      Just (String "Closure") -> pure AKClosure
@@ -512,7 +519,7 @@ instance FromJSON Trait where
 
 instance FromJSON TraitItem where
     parseJSON = withObject "TraitItem" $ \v ->
-                case HML.lookup "kind" v of
+                case lookupKM "kind" v of
                   Just (String "Method") -> do
                     TraitMethod <$> v .: "item_id" <*> v .: "signature"
                   Just (String unk) -> fail $ "unknown trait item type: " ++ unpack unk
@@ -538,5 +545,14 @@ instance FromJSON Static where
            <*> v .: "ty"
            <*> v .: "mutable"
 
-           
+
 --  LocalWords:  initializer supertraits deserialization impls
+
+-- TODO: When the ecosystem widely uses aeson-2.0.0.0 or later, remove this CPP.
+#if MIN_VERSION_aeson(2,0,0)
+lookupKM :: Key -> KM.KeyMap Value -> Maybe Value
+lookupKM = KM.lookup
+#else
+lookupKM :: Text -> HML.HashMap Text Value -> Maybe Value
+lookupKM = HML.lookup
+#endif

--- a/crux/crux.cabal
+++ b/crux/crux.cabal
@@ -21,7 +21,7 @@ library
 
   build-depends:
     base >= 4 && < 5,
-    aeson,
+    aeson < 2.1,
     aeson-typescript,
     aig,
     array,

--- a/crux/src/Crux/SVCOMP.hs
+++ b/crux/src/Crux/SVCOMP.hs
@@ -4,6 +4,7 @@
 --   reporting results, etc.
 
 {-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE DeriveGeneric #-}
 {-# LANGUAGE KindSignatures #-}
@@ -28,8 +29,13 @@ import qualified Data.Yaml as Yaml
 import           GHC.Generics (Generic)
 import           System.FilePath
 import qualified System.FilePath.Glob as Glob
-import qualified Data.HashMap.Strict as HM
 import qualified Data.Vector as V
+
+#if MIN_VERSION_aeson(2,0,0)
+import qualified Data.Aeson.KeyMap as KM
+#else
+import qualified Data.HashMap.Strict as HM
+#endif
 
 import Crux.Config
 import Crux.Config.Common
@@ -227,19 +233,19 @@ loadVerificationTask fp =
  where
  decodeVT :: Yaml.Value -> IO VerificationTask
  decodeVT (Yaml.Object o) =
-  do case HM.lookup "format_version" o of
+  do case lookupKM "format_version" o of
        Just (Yaml.String "2.0") -> return ()
        _ -> fail $ unwords ["Expected verification task version 2.0 while parsing", show fp]
 
-     fs <- case HM.lookup "input_files" o of
+     fs <- case lookupKM "input_files" o of
              Just v -> getStrs v
              Nothing -> fail $ unwords ["No 'input_files' section while parsing", show fp]
 
-     ps <- case HM.lookup "properties" o of
+     ps <- case lookupKM "properties" o of
              Just v -> getProps v
              Nothing -> fail $ unwords ["No 'properties' section while parsing", show fp]
 
-     lang <- case HM.lookup "options" o of
+     lang <- case lookupKM "options" o of
                Just v -> getLang v
                Nothing -> fail $ unwords ["No 'options' section while parsing", show fp]
 
@@ -266,13 +272,13 @@ loadVerificationTask fp =
  getStrs v = fail $ unwords ["expected strings, but got:", show v]
 
  getProp (Yaml.Object o) =
-   do propf <- case HM.lookup "property_file" o of
+   do propf <- case lookupKM "property_file" o of
                  Just (Yaml.String f) -> return (Text.unpack f)
                  _ -> fail $ unwords ["expected string value in 'property_file' while parsing", show fp]
 
       prop <- decodeProp =<< Text.readFile (takeDirectory fp </> propf)
 
-      verdict <- case HM.lookup "expected_verdict" o of
+      verdict <- case lookupKM "expected_verdict" o of
                    Nothing -> return Nothing
                    Just (Yaml.Bool b) -> return (Just b)
                    Just (Yaml.String "true") -> return (Just True)
@@ -287,7 +293,7 @@ loadVerificationTask fp =
  getProps _v = fail $ unwords ["expected property array in 'properties' section when parsing", show fp]
 
  getLang (Yaml.Object o) =
-   case HM.lookup "language" o of
+   case lookupKM "language" o of
      Just (Yaml.String "C")    -> getCDataModel o
      Just (Yaml.String "Java") -> return Java
      _ -> fail $ unwords ["expected 'C' or 'Java' in 'language' while parsing", show fp]
@@ -295,10 +301,17 @@ loadVerificationTask fp =
  getLang _v = fail $ unwords ["expected object in 'options' section when parsing", show fp]
 
  getCDataModel o =
-   case HM.lookup "data_model" o of
+   case lookupKM "data_model" o of
      Just (Yaml.String "ILP32") -> return (C ILP32)
      Just (Yaml.String "LP64")  -> return (C LP64)
      _ -> fail $ unwords ["expected 'ILP32' or 'LP64' in 'data_model' while parsing", show fp]
+
+ -- TODO: When the ecosystem widely uses aeson-2.0.0.0 or later, remove this CPP.
+#if MIN_VERSION_aeson(2,0,0)
+ lookupKM = KM.lookup
+#else
+ lookupKM = HM.lookup
+#endif
 
 compilePats :: [Text] -> [Glob.Pattern]
 compilePats xs =

